### PR TITLE
How 'apply' and 'sync' work

### PR DESCRIPTION
**Pull request by dfellis**
Still a draft, but hopefully this has enough to help everyone understand how the core of IaSQL works and what could potentially trip people up with the module development.

### CODE REVIEW

File: docs/reference/apply_and_sync.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+* Picking the right field for the `@cloudId` in order to both: 1) properly distinguish records in the cloud and 2) still support distinguishing records in the database can sometimes be a difficult job, particularly if the only real candidate is a cloud-generated property.
+* Records that are actually really read-only from the cloud and can't actually be mutated by end users (but are still useful to have to make inserts/updates of other entities safer via join logic) need to have weirdly "backwards" cloud CRUD functions that grab the cloud cache record and re-save it to the database (or delete the user-inserted entity outright).
+* The create-only/create-only-with-unique-columns entities currently require 100% type-unsafe actions to manipulate the DB cache. There are other sections of the codebase where type safety / guarding by the compiler is broken, but this is the worst/most-dangerous one.
```
Comment (aguillenv) [inline docs/reference/apply_and_sync.md:81]: what do you mean with `The create-only/create-only-with-unique-columns entities`? maybe an example here would be good

File: docs/reference/apply_and_sync.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+## Mapper Footguns
+
+This section is meant for anyone currently developing in IaSQL, so it will change much more frequently compared to the sections above.
+
+* TypeORM sometimes violates the entity types you have provided, especially around nullable types, where the entity may be something like `foo?: string` which implies `string | undefined` but TypeORM will treat it as `string | null`. This can trip you up in the `equals` function as `Object.is` distinguishes between `undefined` and `null` so you will run into infinite `update` loops where it is detecting a difference but there is no actionable change to be had. (Similarly sometimes arrays are actually unordered sets, so comparing array element-by-element is incorrect and you need to check each element is inside of the other array, regardless of order, and some strings are actually JSON or other human-editable data formats and non-significant-whitespace differences can also cause `update` loops.
```
[the remark this answers is not in the document]
Comment (yrobla) [reply docs/reference/apply_and_sync.md:77]: indeed, i was seeing a similar problem

File: docs/reference/apply_and_sync.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+The most basic architectural diagram is simply your IaSQL database and your cloud account with the `iasql-engine` in between.
+
+```
+   ______
+  /      \       _______            __
+  |\____/|      |       |       ___/  \__
+  |  DB  | <==> | IaSQL | <==> /  Cloud  \
+  |      |      |_______|      \_    _   /
+  \______/                       \__/ \_/
+```
```
[the remark this answers is not in the document]
Comment (dfellis) [reply docs/reference/apply_and_sync.md:35]: I probably spent too much time making that cloud, but it managed to also get me into "flow" for this document, so maybe that's fine; just woke up late today for it.

File: docs/reference/apply_and_sync.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dfellis) [reply docs/reference/apply_and_sync.md:77]: I didn't want to include our current solutions in this doc because they should change/improve over time until we figure out how to do it in a generalized way and we don't have to do it anymore, then we could just delete the bullet point?

But maybe linking into the codebase where one of the solutions is could be good enough for now.

File: docs/reference/apply_and_sync.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+* Deeply-nested entities get translated into many tables in TypeORM, and unique constraints on those tables can cause lots of pain with the entities that were generated from the cloud APIs, as they don't have the the database-only columns that let TypeORM know when the entity already exists in the database. Unfortunately for now you have to manually find these duplicates and then patch your sub-entity with these DB-only fields or the DB CRUD operations will blow up in your face.
+* Picking the right field for the `@cloudId` in order to both: 1) properly distinguish records in the cloud and 2) still support distinguishing records in the database can sometimes be a difficult job, particularly if the only real candidate is a cloud-generated property.
```
[the remark this answers is not in the document]
Comment (dfellis) [reply docs/reference/apply_and_sync.md:79]: Yes, that should definitely be in here. Though the reason for the different names is that `@cloudId` refers explicitly to an ID that the cloud service is fully aware of while `entityId` is one that we believe we can uniquely create from the data in our *entity* representation but is not "official".

File: docs/reference/apply_and_sync.md (hunk 1 above)
Comment (dfellis) [reply docs/reference/apply_and_sync.md:81]: Oh, this is referring to the security group mapper and ones like it, where you have to `if(ctx?.memo?.db?.<EntityName>?.<IdHere>) ctx.memo.db.<EntityName>.<IdHere> = myReplacementThing;`

100% not typechecked by Typescript so it's always a bit dangerous to do it without extensive integration testing.

File: docs/reference/apply_and_sync.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (aguillenv) [reply docs/reference/apply_and_sync.md:77]: Yes, the lint to the current example in the code seems valid to me.

File: docs/reference/apply_and_sync.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (aguillenv) [reply docs/reference/apply_and_sync.md:79]: Hmm but `cloudId` overrides the `entityId`, right?

File: docs/reference/apply_and_sync.md (hunk 1 above)
Comment (aguillenv) [reply docs/reference/apply_and_sync.md:81]: ohh I see. Same as before, maybe a link to a place in the code where this is being done for context would be good!

**Review by aguillenv — APPROVED** [review]

File: docs/reference/apply_and_sync.md (hunk 1 above)
Comment (dfellis) [reply docs/reference/apply_and_sync.md:81]:
```suggestion
* The create-only/create-only-with-unique-columns entities currently require [100% type-unsafe actions to manipulate the DB cache](https://github.com/iasql/iasql-engine/blob/main/src/modules/0.0.15/aws_security_group/index.ts#L171). There are other sections of the codebase where type safety / guarding by the compiler is broken, but this is the worst/most-dangerous one.
```

File: docs/reference/apply_and_sync.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dfellis) [reply docs/reference/apply_and_sync.md:77]:
```suggestion
* TypeORM sometimes violates the entity types you have provided, especially around nullable types, where the entity may be something like `foo?: string` which implies `string | undefined` but TypeORM will treat it as `string | null`. This can trip you up in the `equals` function as `Object.is` distinguishes between `undefined` and `null` so you will run into infinite `update` loops where it is detecting a difference but there is no actionable change to be had. [We currently re-adjust TypeORM output after to read or writes to the database to avoid this](https://github.com/iasql/iasql-engine/blob/main/src/modules/0.0.15/aws_security_group/entity/index.ts#L51-L59). (Similarly sometimes arrays are actually unordered sets, so comparing array element-by-element is incorrect and you need to check each element is inside of the other array, regardless of order, and some strings are actually JSON or other human-editable data formats and non-significant-whitespace differences can also cause `update` loops.
```

File: docs/reference/apply_and_sync.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+This dance is necessary to give users in IaSQL the ability to migrate from *any* cloud state to the state they have defined in their database. This is IaSQL's superpower: bringing a greater reliability to producing the cloud state the developer specifies, so they can worry about more important things.
+
+## Mapper Footguns
```
Comment (depombo) [inline docs/reference/apply_and_sync.md:73]: This should go in CONTRIBUTING.md and not here?

File: docs/reference/apply_and_sync.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
...it first sees that it is supposed to mutate the security group description without mutating the name but alters the name in its cache and executes a cloud create with this new, fake record. Then the inner loop runs and entities connected to the security group "realize" they are on the wrong security group and switch to this new, temporary security group, while the old security group with the correct name is marked as cloud-only so it is deleted. Then the inner loop completes with no difference between the database *cache* and the cloud, so it goes back to the outer loop, which re-loads the *actual* database state and it checks again and sees that there's a security group in the database that is not in the cloud and a security group in the cloud that is not in the database, and that these other entities are connected to the wrong security group, so it creates tasks to create a security group, delete a security group, and switch security groups. The error-retrying-loop will execute these tasks some unknown number of times, but the successes will always be in the same order -> create security group -> switch other entities to that security group -> delete the temporary security group.
```
Comment (depombo) [inline docs/reference/apply_and_sync.md:69]: This loop will be hard to follow for someone coming it. We have the security group example, but it might be a good idea to re-state at the end what the user does exactly (SQL included?), what the loop does under the hood and what end/goal state is achieved?

**Comment by depombo** [comment]
This has already been ported over to the docs' new location
